risc0/zkvm/src/host/protos/api.proto

```diff
@@ -0,0 +1,240 @@
+syntax = "proto3";
+
+import "google/protobuf/empty.proto";
+
+package protos.api;
+
+message ServerRequest {
+  oneof kind {
+    ExecuteRequest execute = 1;
+    ProveRequest prove = 2;
+    ProveSegmentRequest prove_segment = 3;
+    LiftRequest lift = 4;
+    JoinRequest join = 5;
+  }
+}
+
+message HelloRequest {
+  uint32 version = 1;
+}
+
+message HelloReply {
+  oneof kind {
+    HelloResult ok = 1;
+    GenericError error = 2;
+  }
+}
+
+message HelloResult {
+  uint32 version = 2;
+}
+
+message ExecuteRequest {
+  ExecutorEnv env = 1;
+  AssetRequest segments_out = 2;
+}
+
+message ProveRequest {
+  ExecutorEnv env = 1;
+  ProverOpts opts = 2;
+  AssetRequest receipt_out = 3;
+}
+
+message ProveSegmentRequest {
+  ProverOpts opts = 1;
+  Asset segment = 2;
+  AssetRequest receipt_out = 3;
+}
+
+message ProveSegmentReply {
+  oneof kind {
+    ProveSegmentResult ok = 1;
+    GenericError error = 2;
+  }
+}
+
+message LiftRequest {
+  ProverOpts opts = 1;
+  Asset receipt = 2;
+  AssetRequest receipt_out = 3;
+}
+
+message LiftReply {
+  oneof kind {
+    LiftResult ok = 1;
+    GenericError error = 2;
+  }
+}
+
+message LiftResult {
+  Asset receipt = 1;
+}
+
+message JoinRequest {
+  ProverOpts opts = 1;
+  Asset left = 2;
+  Asset right = 3;
+  AssetRequest receipt_out = 4;
+}
+
+message JoinReply {
+  oneof kind {
+    JoinResult ok = 1;
+    GenericError error = 2;
+  }
+}
+
+message JoinResult {
+  Asset receipt = 1;
+}
+
+message ExecutorEnv {
+  Binary binary = 1;
+  map<string, string> env_vars = 2;
+  repeated string slice_ios = 3;
+  repeated uint32 read_fds = 4;
+  repeated uint32 write_fds = 5;
+}
+
+message Binary {
+  enum Kind {
+    UNSPECIFIED = 0;
+    IMAGE = 1;
+    ELF = 2;
+  }
+
+  Kind kind = 1;
+  Asset asset = 2;
+}
+
+message ProverOpts {
+  string hashfn = 1;
+  bool lift = 2;
+}
+
+message SessionInfo {
+  uint32 segments = 1;
+  bytes journal = 2;
+  ExitCode exit_code = 3;
+}
+
+message SegmentInfo {
+  uint32 index = 1;
+  uint32 po2 = 2;
+  uint32 insn_cycles = 3;
+  Asset segment = 4;
+}
+
+message ProveSegmentResult {
+  Asset receipt = 1;
+}
+
+message ExitCode {
+  oneof kind {
+    uint32 halted = 1;
+    uint32 paused = 2;
+    google.protobuf.Empty system_split = 3;
+    google.protobuf.Empty session_limit = 4;
+  }
+}
+
+message Asset {
+  oneof kind {
+    bytes inline = 1;
+    string path = 2;
```

Comment:
Is this `path` intended to be created and managed by the server, or is the client intended to be able to create an asset as a file (or S3 blob?) and then hand the server a path which it will read? If these paths are only created by the server, maybe we should treat them as opaque identifiers from the API point of view.

Reply:
The `AssetRequest` is intended for clients to fill in the base path where they'd like assets to go. In the case of EFS, we have a mount that the client knows about and we're asking the server to ensure that assets show up there.

Reply:
Ah ok that makes sense. I mentioned it because having the client direct the server to write a file to an arbitrary path feels dangerous if they are not on the same machine with the same privileges, but my understanding is that this revision of the API is only intended to be used in a single-machine context.

Reply:
I've been concerned about the security ramifications too. In this case it's fine, but we have to be careful with it in other scenarios. Alternatively, we could make this a `--root_dir` argument to `r0vm` or something rather than allowing clients to set the base path in-band. This would also nicely restrict it to the sub-process case where permissions are reasonable since they inherit from the parent.